src/components/CodeMirrorEditor.jsx

```diff
@@ -0,0 +1,90 @@
+import CodeMirror from 'codemirror';
+import isNil from 'lodash-es/isNil';
+import LRU from 'lru-cache';
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+import 'codemirror/mode/htmlmixed/htmlmixed';
+import 'codemirror/mode/css/css';
+import 'codemirror/mode/javascript/javascript';
+
+const CODEMIRROR_MODES_MAP = {
+  html: 'htmlmixed',
+  css: 'css',
+  javascript: 'javascript',
+};
+
+export default function CodeMirrorEditor({
+  language,
+  projectKey,
+  source,
+  onInput,
+}) {
+  const containerRef = useRef();
+  const editorRef = useRef();
```

Comment:
[The docs](https://reactjs.org/docs/hooks-reference.html#useref) show that you're supposed to provide an initial value- I would suggest doing `useRef(null)` to make this clear

Reply:
The docs don't make a particularly strong claim that the initial value is required, I don't think, and [this example from the Hooks FAQ](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables) shows `useRef()` called without an argument, so I think it's legit. And semantically it makes sense, just like a `let` that declares a variable without assigning it, thus giving it an `undefined` initial value.